src/librustc/infer/error_reporting/mod.rs

```diff
@@ -1067,6 +1067,43 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                                sub_region: Region<'tcx>,
                                sup_origin: SubregionOrigin<'tcx>,
                                sup_region: Region<'tcx>) {
+
+        // #45983: when trying to assign the contents of an argument to a binding outside of a
+        // closure, provide a specific message pointing this out.
+        if let (&SubregionOrigin::BindingTypeIsNotValidAtDecl(ref external_span),
+                &RegionKind::ReFree(ref free_region)) = (&sub_origin, sup_region) {
+            let hir = &self.tcx.hir;
+            if let Some(node_id) = hir.as_local_node_id(free_region.scope) {
+                match hir.get(node_id) {
+                    hir_map::NodeExpr(hir::Expr {
+                        node: hir::ExprClosure(_, _, _, closure_span, false),
+                        ..
+                    }) => {
+                        let sup_sp = sup_origin.span();
+                        let origin_sp = var_origin.span();
+                        let mut err = self.tcx.sess.struct_span_err(
+                            sup_sp,
+                            "borrowed data cannot be moved outside of its closure");
+                        err.span_label(sup_sp, "cannot be moved outside of its closure");
+                        if sup_sp == origin_sp {
+                            err.span_label(*external_span,
+                                           "borrowed data cannot be moved into here...");
```

Comment:
I still find the word `move` potentially misleading. I would prefer "stored".

Reply:
Changed